# This patch is a fix for #2111.
# It ensures that shuffle is lowered for this particular case correctly.
diff --git a/llvm/lib/Target/X86/X86ISelLowering.cpp b/llvm/lib/Target/X86/X86ISelLowering.cpp
index 6b816c710f98..b26fda666acc 100644
--- a/llvm/lib/Target/X86/X86ISelLowering.cpp
+++ b/llvm/lib/Target/X86/X86ISelLowering.cpp
@@ -43192,9 +43192,10 @@ static SDValue combineHorizOpWithShuffle(SDNode *N, SelectionDAG &DAG,
           ShuffleVectorSDNode::commuteMask(ShuffleMask1);
         }
         if ((Op00 == Op10) && (Op01 == Op11)) {
-          SmallVector<int, 4> ShuffleMask;
-          ShuffleMask.append(ShuffleMask0.begin(), ShuffleMask0.end());
-          ShuffleMask.append(ShuffleMask1.begin(), ShuffleMask1.end());
+          const int Map[4] = {0, 2, 1, 3};
+          SmallVector<int, 4> ShuffleMask(
+            {Map[ShuffleMask0[0]], Map[ShuffleMask1[0]], Map[ShuffleMask0[1]],
+             Map[ShuffleMask1[1]]});
           SDLoc DL(N);
           MVT ShufVT = VT.isFloatingPoint() ? MVT::v4f64 : MVT::v4i64;
           SDValue Res = DAG.getNode(Opcode, DL, VT, Op00, Op01);
